### PR TITLE
BUG: Fix spectral extraction in the presence of unit conversion (no actual conversion, no new callback)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is an alternative to https://github.com/spacetelescope/jdaviz/pull/3168 . This patch avoid actually converting the spectrum. And it also avoids adding yet another listener (fear of further performance impact). It targets only where this particular bug is happening in the `add_results_to_plugin` method.

Close https://github.com/spacetelescope/jdaviz/pull/3168

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-4734)

### CODE REVIEW

**Comment by kecnry** [comment]
This solution is isolated to the yunit for spectral extraction in cubeviz, but the root problem is much more general. Do we want to just fix this now and open a ticket to replace with the general solution later?

**Comment by pllim** [comment]
> Do we want to just fix this now and open a ticket to replace with the general solution later?

That would be my preference, yes. I don't want to fix a problem that does not yet exist because I cannot tell what would be the best fix right now, especially with other performance issues getting addressed.

**Comment by pllim** [comment]
Also interesting to note that the X-axis is not affected by this problem at all.

**Comment by kecnry** [comment]
>  I don't want to fix a problem that does not yet exist

Can this not currently also be reproduced in other plugins or when unloading/loading data into a viewer?  If so, then I definitely agree, but I suspect the reported case is just one case of a more general bug.

**Comment by pllim** [comment]
> other plugins

What other plugins removes and re-adds data? I cannot think of any.

> when unloading/loading data into a viewer

I thought it was debatable whether behavior in this case is truly a bug or not?

**Comment by kecnry** [comment]
> What other plugins removes and re-adds data? I cannot think of any

Any that allow replacing a current entry, so for 1d spectra from cubeviz that probably leaves model fitting and gaussian smooth.

> I thought it was debatable whether behavior in this case is truly a bug or not?

Ok, then let's discuss that.  I think it would be confusing if you unload all data from a viewer, start adding them back in but then the user-selected display unit, which are intended to be app-wide/global are no longer respected until you toggle the selection in the unit conversion plugin and then they resync.

**Comment by gibsongreen** [comment]
With the most recent commit, if I model fit from a spectrum, and the display units are not in the native data collection units, should the model be in the native units or the current display units?

**Comment by pllim** [comment]
> should the model be in the native units or the current display units?

I think that is a separate ticket? I don't remember if we reached a consensus. Maybe @kecnry knows.

**Comment by kecnry** [comment]
@gibsongreen - does this change the behavior from main (I don't see where in the code that would be affected)?  I'm actually not sure what I'd expect, ultimately I think our goal should be to minimize compounding errors... so in this case it may make sense for the model to be in the current display units since that is what the input parameters are provided in (although the data is in native units, so I can see an argument either way).

**Comment by pllim** [comment]
Should I exclude model fitting from the patch to be less controversial?

**Comment by gibsongreen** [comment]
> @gibsongreen - does this change the behavior from main

It does appear to have the same behavior as main (model item in data collection loading in with the current display unit). I'm not sure either on what exactly to expect, it makes sense to me that the model units are stored as the display units. I also think that extracted spectrum should be the native units (even after removed and then re-extracted) since that how it works with the original extracted spectrum. @pllim so if we are good with this I wouldn't think this particular comment needs a follow up. Everything else seems good to me so whatever we think is best here.

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py [hunk 1]
```diff
@@ -569,6 +570,17 @@ def test_default_spectral_extraction(cubeviz_helper, spectrum1d_cube_fluxunit_jy
...
+    # Also test unit conversion
+    uc = cubeviz_helper.plugins["Unit Conversion"]
+    assert uc.flux_unit == "Jy"
+    uc.flux_unit = "MJy"
+    spec_extr_plugin = cubeviz_helper.plugins['Spectral Extraction']
+    collapsed = spec_extr_plugin.extract()
+    # Actual values not in display unit but should not affect display unit.
+    assert collapsed.flux.unit == u.Jy
+    assert uc.flux_unit.selected == "MJy"
+    assert cubeviz_helper.app._get_display_unit('spectral_y') == "MJy"
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py:582]: This actually does not fail on `main` even though I see the bug when I re-extract the one spectrum manually. Not sure how to make this test actually catch that bug. 🤷‍♀️

File: jdaviz/configs/specviz/tests/test_viewers.py [hunk 2]
```diff
@@ -26,6 +26,21 @@ def test_spectrum_viewer_axis_labels(specviz_helper, input_unit, y_axis_label):
     assert (y_axis_label in label)
 
 
+@pytest.mark.xfail(reason="FIXME: Some callback magic needs to happen somewhere.")
```
Comment (pllim) [inline jdaviz/configs/specviz/tests/test_viewers.py:29]: @kecnry , I think this is the more general case you wanted me to fix? My callback-fu isn't strong enough but I added this test here and mark it xfail. If you do follow-up ticket, you can remove the xfail. Thanks.

Comment (kecnry) [reply jdaviz/configs/specviz/tests/test_viewers.py:29]: does this currently fail?  We might need to also assert that the actual units on the viewer match, I'm not sure

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py:582]: the last two lines are effectively doing the same thing.  If the viewer is unsyncing from the unit conversion plugin, then we'll need to check the actual units on the viewer (`viewer.state.x_display_unit` or something like that).

File: jdaviz/configs/specviz/tests/test_viewers.py (hunk 2 above)
Comment (pllim) [reply jdaviz/configs/specviz/tests/test_viewers.py:29]: It fails on main and with this patch, yes.

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py:582]: But I did confirm manually that this patch should fix the original problem Clare reported.

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py:582]: Oh, oops. I didn't see there is a subset in  that test. I moved the test code to a new function and it fails on main but passes here. I will push new commit here.

**Review by gibsongreen — APPROVED** [review]
These changes look good to me, thanks!

**Pull request merged**